File: gcf/test/test_cs/gcf/Tab/Parser.cs
namespace Tab
{
    public class Parser
    {
        class TokenStream : Serialize.ITokenReadStream
        {
            Parser _parser = null;
            public TokenStream(Parser parser)
            { _parser = parser; }
            public string Read()
            { return _parser.Pop(); }
        }

        IndexMap _idxMap;
        TokenStream _stream;
        Serialize.TextReader _reader;
        int _index = 0;
        string[] _data = null;

        public Parser(IndexMap idxMap)
        {
            _idxMap = idxMap;
            _stream = new TokenStream(this);
            _reader = new Serialize.TextReader(_stream);
        }

        public Serialize.IReader Parse(string line)
        {
            if (string.IsNullOrEmpty(line))
                return null;

            _index = 0;
            _data = line.Split('\t');
            return _reader;
        }

        string Pop()
        {
            if (_data == null)
                return null;

            int idx = _idxMap.Index(_index);
            ++_index;

            if (idx == -1 || idx >= _data.Length)
                return null;

            return _data[idx];
        }
    }
}

File: gcf/test/gcf/src/gcf/tab/Parser.cpp
#include "Parser.h"
#include <cassert>

namespace tab
{
    Parser::Parser(const IndexMap& idxMap)
        : _idxMap(idxMap)
        , _stream(*this)
        , _reader(&_stream)
    {
    }

    Parser::~Parser()
    {
    }

    serialize::IReader* Parser::Parse(char* line)
    {
        int idx = 0;
        int beg = 0;

        _tokens.assign(_idxMap.TitleSize(), "");
        while (idx < (int)_idxMap.TitleSize())
        {
            int pos = beg;
            while (line[pos] != 0 && line[pos] != '\t')
                ++pos;

            _tokens[idx++] = line + beg;

            if (line[pos] == 0)
                break;

            line[pos] = 0;
            beg = pos + 1;
        }

        _idx = 0;
        return &_reader;
    }

    const char* Parser::Pop()
    {
        if (_idx < 0 || _idx >= (int)_idxMap.Size())
            return nullptr;

        int realIdx = _idxMap.Index(_idx++);
        if (realIdx == -1)
            return "";

        assert(realIdx <= _tokens.size());
        return _tokens[realIdx];
    }

    bool Parser::Push(const char*)
    {
        assert(false);
        return false;
    }
}
